### ISSUE TITLE
make returning errors consistent

### ISSUE THREAD

**Issue by tmakatos** [nutanix/libvfio-user#80]
Some functions return -1 and set `errno`, others returns `-errno`. We should be consistent, e.g. make all functions return `-errno`.

**Comment by jlevon** [comment]
I'm not quite following the need for this -errno stuff. Usually, this is done - in the form of PTR_ERR() and friends - when we're returning a pointer, but we are doing this all over the place not just for that instance. There shouldn't be any need to do this via a "0 is success, otherwise you get an errno". There's quite a bit of code that finds itself switching the signs and it's a little hard to follow.

**Comment by tmakatos** [comment]
Is `PTR_ERR` and friends acceptable to use in user space? What we want to achieve here is to have a consistent way across functions  for returning error codes, I'm not particular how we do it, I'm open to suggestions.

**Comment by jlevon** [comment]
I haven't seen it in userspace before TBH.

So far, there isn't really that much code that returns pointers, is there?
So I'd be inclined to not do the negative thing overall. Agreed on consistency for sure!

I would be inclined to return "errno" - so 0 for success, otherwise ENOENT etc.
Does that sound reasonable? Another preference?

**Comment by swapnili** [comment]
But if we return,
0   - Success
>0 - Failure

Then this will not address the case where function want to return success non-zero value e.g size
I think -errno solves this issue. So
>=0 - Success
<0   -  Failure

**Comment by jlevon** [comment]
That's a good point. But I'm also not sure we have too many of those. The int get_foo(size_t *sizep) would work for those, but it's a bit less elegant.

I really don't think -errno is appropriate for the public API - I don't know offhand of any other public APIs that do this. And given that, and how many places we'd have to be careful about public versus internal, I think it's best to stick with one of the standard C-library methods.

Having said that, if both you and Thanos think we should do -errno even for the public API we can go with that.

**Comment by swapnili** [comment]
No I did not meant to say this behaviour for public API's. I meant it for internal functions.  I thought this issue was opened for internal functions and not public API's.

**Comment by tmakatos** [comment]
I was planning to address both internal and public APIs with this issue. We've been returing -errno simply because this is how the kernel does it. The benefit of this approach is that, as Swapnil said, we can return 0 for success and >0 for values. TBH I don't know whether there is a standard way public libraries APIs solve this.

Let's enumerate our options (we can also state pros/cons for each option):

1. Return -errno for error, >= 0 for success/value.
2. Return -1 for error, set `errno`.
3. Use PTR_ERR/ERR_PTR.
4. Pass pointer to function to communicate error.

Personally I find option 1 the simplest and least invasive in the code. Shall we have a poll on Slack? :)

**Comment by jlevon** [comment]
I'm really loathe to set a new precedent for the public API here

**Comment by jlevon** [comment]
I just found another bug due to getting this inconsistent.
I'm going to pick up cleaning this up. In the interests of making progress, I'm going to just pick my personal preference here, which is the traditional userspace model: setting errno, and returning -1 or NULL, on error.

**Comment by jlevon** [comment]
Fixed by series ending in #433
